lib/internal/test_runner/tests_stream.js

```diff
@@ -0,0 +1,74 @@
+'use strict';
+const {
+  ArrayPrototypePush,
+  ArrayPrototypeShift,
+} = primordials;
+const Readable = require('internal/streams/readable');
+
+class TestsStream extends Readable {
+  #buffer;
+  #canPush;
+
+  constructor() {
+    super({ objectMode: true });
+    this.#buffer = [];
+    this.#canPush = true;
+  }
+
+  _read() {
+    this.#canPush = true;
+
+    while (this.#buffer.length > 0) {
+      const obj = ArrayPrototypeShift(this.#buffer);
+
+      if (!this.#tryPush(obj)) {
+        return;
+      }
+    }
+  }
+
+  fail(nesting, testNumber, name, details, directive) {
+    this.#emit('test:fail', { __proto__: null, name, nesting, testNumber, details, ...directive });
+  }
+
+  ok(nesting, testNumber, name, details, directive) {
+    this.#emit('test:pass', { __proto__: null, name, nesting, testNumber, details, ...directive });
+  }
+
+  plan(nesting, count) {
+    this.#emit('test:plan', { __proto__: null, nesting, count });
+  }
+
+  getSkip(reason) {
```

Comment:
nit: marking the param explicitly as optional helps with code readability IMO.
```suggestion
  getSkip(reason = undefined) {
```